### PR TITLE
[lldb] Avoid (unlimited) GetNumChildren calls when printing values

### PR DESCRIPTION
**Pull request by labath**
For some data formatters, even getting the number of children can be an expensive operations (e.g., needing to walk a linked list to determine the number of elements). This is then wasted work when we know we will be printing only small number of them.

This patch replaces the calls to GetNumChildren (at least those on the "frame var" path) with the calls to the capped version, passing the value of `max-children-count` setting (plus one)

### CODE REVIEW

**Review by JDevlieghere — APPROVED** [review]

**Comment by jimingham** [comment]
This seems reasonable.  However, I note that on the page where we show how to write synthetic child providers, we say:

   def num_children(self):
      this call should return the number of children that you want your object to have

That's actually not true - we pass the max_children argument on to the num_children method, and in fact some of the tests do use the max parameter.  But since you're making that actually useful, can you fix the docs so people will know to take advantage of this?

**Comment by jimingham** [comment]
BTW, because the max number of children is probably getting ignored in the wild, this has to be advisory, you can't require that:

val.GetNumChildren(max_children) <= max_children

I don't think you do that but it might be good to note the fact.

File: lldb/source/DataFormatters/FormatManager.cpp [hunk 1]
```diff
@@ -442,16 +443,19 @@ lldb::Format FormatManager::GetSingleItemFormat(lldb::Format vector_format) {
...
-      !valobj.GetTargetSP()->GetDebugger().GetAutoOneLineSummaries())
+  if (target_sp && !target_sp->GetDebugger().GetAutoOneLineSummaries())
     return false; // then don't oneline
 
   // if this object has a summary, then ask the summary
   if (valobj.GetSummaryFormat().get() != nullptr)
     return valobj.GetSummaryFormat()->IsOneLiner();
 
-  auto num_children = valobj.GetNumChildren();
+  size_t max_num_children =
```
[the remark this answers is not in the document]
Comment (labath) [reply lldb/source/DataFormatters/FormatManager.cpp]: Done

**Comment by labath** [comment]
> This seems reasonable. However, I note that on the page where we show how to write synthetic child providers, we say:
> 
> def num_children(self): this call should return the number of children that you want your object to have
> 
> That's actually not true - we pass the max_children argument on to the num_children method, and in fact some of the tests do use the max parameter. But since you're making that actually useful, can you fix the docs so people will know to take advantage of this?

Sounds good. I've created a separate [pr](https://github.com/llvm/llvm-project/pull/94192) for that.


> BTW, because the max number of children is probably getting ignored in the wild, this has to be advisory, you can't require that:
> 
> val.GetNumChildren(max_children) <= max_children
> 
> I don't think you do that but it might be good to note the fact.

Yes, I am keeping that in mind.

**Pull request merged**